x-pack/plugins/cases/server/services/user_actions/transform.ts

```diff
@@ -0,0 +1,316 @@
+      actionFields,
+      actionDetails: decodedJson,
+      fieldType,
+    });
+
+    return { transformedActionDetails, references };
+  } catch (error) {
+    return defResult;
+  }
+}
+
+export function extractConnectorId({
+  action,
+  actionFields,
+  actionDetails,
+  fieldType,
+}: {
+  action: string;
+  actionFields: string[];
+  actionDetails?: Record<string, unknown> | string | null;
+  fieldType: UserActionFieldType;
+}): {
+  transformedActionDetails?: string | null;
+  references: SavedObjectReference[];
+} {
+  if (!actionDetails || isString(actionDetails)) {
+    // the action was null, undefined, or a regular string so just return it unmodified and not encoded
+    return { transformedActionDetails: actionDetails, references: [] };
+  }
+
+  try {
+    return extractConnectorIdHelper({
+      action,
+      actionFields,
+      actionDetails,
+      fieldType,
+    });
+  } catch (error) {
+    return { transformedActionDetails: encodeActionDetails(actionDetails), references: [] };
+  }
+}
+
+function encodeActionDetails(actionDetails: Record<string, unknown>): string | null {
+  try {
+    return JSON.stringify(actionDetails);
+  } catch (error) {
+    return null;
+  }
+}
+
+export function extractConnectorIdHelper({
+  action,
+  actionFields,
+  actionDetails,
+  fieldType,
+}: {
+  action: string;
+  actionFields: string[];
+  actionDetails: unknown;
+  fieldType: UserActionFieldType;
+}): { transformedActionDetails: string; references: SavedObjectReference[] } {
+  let transformedActionDetails: unknown = actionDetails;
+  let references: SavedObjectReference[] = [];
+
+  try {
+    if (isCreateCaseConnector(action, actionFields, actionDetails)) {
+      ({ transformedActionDetails, references } = transformCreateConnector(
+        actionDetails.connector,
+        fieldType
+      ));
+    } else if (isUpdateCaseConnector(action, actionFields, actionDetails)) {
+      ({ transformedActionDetails, references } = transformUpdateConnector(
+        actionDetails,
+        fieldType
+      ));
+    } else if (isPushConnector(action, actionFields, actionDetails)) {
+      ({ transformedActionDetails, references } = transformPushConnector(actionDetails, fieldType));
+    }
+  } catch (error) {
+    // ignore any errors, we'll just return whatever was passed in for action details in that case
+  }
+
+  return { transformedActionDetails: JSON.stringify(transformedActionDetails), references };
+}
+
+function isCreateCaseConnector(
+  action: string,
+  actionFields: string[],
+  actionDetails: unknown
+): actionDetails is { connector: CaseConnector } {
+  try {
+    const unsafeCase = actionDetails as CaseAttributes;
+
+    return (
+      isCreateConnector(action, actionFields) &&
+      unsafeCase.connector !== undefined &&
+      CaseConnectorRt.is(unsafeCase.connector)
+    );
+  } catch {
+    return false;
+  }
+}
+
+export const ConnectorIdReferenceName: Record<UserActionFieldType, ConnectorIdRefNameType> = {
```

Comment:
A map to quickly grab the reference name we'll use in the saved object reference depending on the field (whether `newValue` or `oldValue`). The names need to be different because it's possible to have both `newValue` and `oldValue` within the references array.